include/irr/asset/ITexturePacker.h

```diff
@@ -0,0 +1,483 @@
+                auto& region = (*regions)[i];
+                region.bufferOffset = bufOffset;
+                region.bufferImageHeight = 0u;
+                region.bufferRowLength = tilesPerLodDim*texelSz;
+                region.imageExtent = {tilesPerLodDim,tilesPerLodDim,1u};
+                region.imageOffset = {0,0,0};
+                region.imageSubresource.baseArrayLayer = 0u;
+                region.imageSubresource.layerCount = 1u;
+                region.imageSubresource.mipLevel = i;
+                region.imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+
+                bufOffset += regionSz;
+            }
+            auto buf = core::make_smart_refctd_ptr<ICPUBuffer>(bufOffset);
+            uint32_t* bufptr = reinterpret_cast<uint32_t*>(buf->getPointer());
+            std::fill(bufptr, bufptr+bufOffset/4u, SPhysPgOffset::invalid_addr);
+
+            m_pageTable->setBufferAndRegions(std::move(buf), regions);
+        }
+        const uint32_t pgSzLog2 = core::findMSB(m_pgSzxy);
+        bool ok = SMiptailPacker::computeMiptailOffsets(m_miptailOffsets, pgSzLog2, TILE_PADDING);
+        assert(ok);
+    }
+
+    void free(page_tab_offset_t _addr, const IImage* _img, const IImage::SSubresourceRange& _subres) override
+    {
+        //free physical pages
+        auto extent = _img->getCreationParameters().extent;
+        const uint32_t levelCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+
+        for (uint32_t i = 0u; i < levelCount; ++i)
+        {
+            const uint32_t w = (std::max(extent.width>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
+            const uint32_t h = (std::max(extent.height>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
+
+            uint32_t* const pgTab = reinterpret_cast<uint32_t*>(
+                reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer()) + m_pageTable->getRegions().begin()[i].bufferOffset
+                );
+            const uint32_t pgtPitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
+            const uint32_t offset = (_addr.y>>i)*pgtPitch + (_addr.x>>i);
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    const uint32_t linTexelAddr = offset + y*pgtPitch + x;
+
+                    SPhysPgOffset physPgOffset = pgTab[linTexelAddr];
+                    if (physPgOffset_valid(physPgOffset))
+                    {
+                        pgTab[linTexelAddr] = SPhysPgOffset::invalid_addr;
+
+                        uint32_t addrs[2] { physPgOffset.addr&0xffffu, physPgOffset_mipTailAddr(physPgOffset).addr&0xffffu };
+                        const uint32_t szs[2]{ 1u,1u };
+                        core::address_allocator_traits<phys_pg_addr_alctr_t>::multi_free_addr(m_physPgAddrAlctr, physPgOffset_hasMipTailAddr(physPgOffset) ? 2u : 1u, addrs, szs);
```

Comment:
`multi_free` can be expensive, better that you buffer up the addresses to be freed